scripts/question_answering/data_pipeline.py

```diff
@@ -0,0 +1,900 @@
+        with open(os.path.join(self._data_root_path, self._processed_train_data_file_name),
+                  'r') as f:
+            train_examples = json.load(f)
+
+        with open(os.path.join(self._data_root_path, self._processed_dev_data_file_name), 'r') as f:
+            dev_examples = json.load(f)
+
+        word_vocab = pickle.load(
+            open(os.path.join(self._data_root_path, self._word_vocab_file_name), 'rb'))
+
+        char_vocab = pickle.load(
+            open(os.path.join(self._data_root_path, self._char_vocab_file_name), 'rb'))
+
+        return train_examples, dev_examples, word_vocab, char_vocab
+
+    def _save_processed_data(self, train_examples, dev_examples, word_vocab, char_vocab):
+        """Save processed data to disk
+
+        Parameters
+        ----------
+        train_examples : List[Tuple]
+            Processed train examples. Each tuple consists of question_id, record_index,
+            context_tokens_indices, question_tokens_indices, context_chars_indices,
+            question_char_indices, start_token_index_of_the_answer, end_token_index_of_the_answer,
+            context, context_tokens_spans
+        dev_examples : List[Tuple]
+            Processed dev examples. Each tuple consists of question_id, record_index,
+            context_tokens_indices, question_tokens_indices, context_chars_indices,
+            question_char_indices, start_token_index_of_the_answer, end_token_index_of_the_answer,
+            context, context_tokens_spans
+        word_vocab : Vocab
+            Word-level vocabulary
+        char_vocab : Vocab
+            Char-level vocabulary
+        """
+        with open(os.path.join(self._data_root_path, self._processed_train_data_file_name),
+                  'w') as f:
+            json.dump(train_examples, f)
+
+        with open(os.path.join(self._data_root_path, self._processed_dev_data_file_name), 'w') as f:
+            json.dump(dev_examples, f)
+
+        pickle.dump(word_vocab,
+                    open(os.path.join(self._data_root_path, self._word_vocab_file_name), 'wb'))
+
+        pickle.dump(char_vocab,
+                    open(os.path.join(self._data_root_path, self._char_vocab_file_name), 'wb'))
+
+    @staticmethod
+    def _partition(mapped_values):
+        """Groups items with same keys into a single partition
+
+        Parameters
+        ----------
+        mapped_values : List[Tuple]
+            List of mapped (key, value) tuples
+
+        Returns
+        -------
+        items: List[Tuple]
+            List of partitions, where each partition is (key, List[value])
+        """
+        partitioned_data = collections.defaultdict(list)
+
+        for key, value in mapped_values:
+            partitioned_data[key].append(value)
+
+        return partitioned_data.items()
+
+
+class SQuADDataTokenizer:
+    """SQuAD data tokenizer, that encapsulate the splitting logic of each entry of SQuAD dataset"""
+    tokenizer = spacy.blank('en')
```

Comment:
only initialize when use_spacy is True. also, is it applicable to use spacy and nltk integration in gluonnlp?

Reply:
I can and will use SpacyTokenizer from gluonnlp, but:

1. I cannot instantiate it only when use_spacy is True, because, when I am doing multiprocessing, the instantiation won't happen as `__init__` method is not called. If you have any thought how to fix it, would gladly apply.

2. I am not able to use `SacreMosesTokenizer` from GluonNLP, because it encodes things like apostrophes by default. That happens because the underlying`sacremoses.MosesTokenizer.tokenize` has a default `escape=True`, and there is no way to turn it to False with existing GLuonNLP API. I consider this as a bug, and we may want to address it